### PR TITLE
Pregnant signups

### PR DESCRIPTION
**Pull request by marcus-a-davis**

### CODE REVIEW

File: management/models.py [hunk 1]
```diff
@@ -43,6 +43,8 @@ class Contact(models.Model):
     org_sign_up = models.CharField(max_length=40, blank=True)
     hospital_name = models.CharField(max_length=50, blank=True)
     doctor_name = models.CharField(max_length=30, blank=True)
+    preg_signup = models.NullBooleanField(default=False)
+    preg_update = models.NullBooleanField(default=False)
```
Comment (peterhurford) [inline management/models.py]: We should document all these fields, I don't think I really understand what they are or what they do.

File: modules/i18n.py [hunk 2]
```diff
@@ -50,11 +51,14 @@ def hindi_remind():
 def hindi_information():
     return u'\u0907\u0924\u094d\u0924\u093f\u0932\u093e'
 
+def hindi_born():
+    return u'\u091c\u0928\u094d\u092e'
+
 def subscribe_keywords(language):
     if language == "English":
-        return ["remind", "join"]
+        return ["remind", "join","born"]
```
Comment (peterhurford) [inline modules/i18n.py]: Space after comma

File: modules/i18n.py [hunk 3]
```diff
@@ -129,3 +133,12 @@ def five_year_reminder_seven_days(language):
...
+    elif language == "Hindi":
+        return hindi_preg_verify_msg()
+
+def hindi_preg_verify_msg():
+    return u'\u0905\u092a\u0928\u0940 \u0917\u0930\u094d\u092d\u093e\u0935\u0938\u094d\u0920\u093e \u0915\u0947 \u0926\u094c\u0930\u093e\u0928 \u0906\u092a\u0928\u0947 \u0938\u094d\u0935\u093e\u0938\u094d\u0925\u094d\u092f \u0938\u094d\u092e\u0930\u0923 \u0915\u0940 \u0938\u0926\u0938\u094d\u092f\u0924\u093e \u0932\u0940 \u0925\u0940. \u092f\u0926\u093f \u0906\u092a\u0915\u093e \u092c\u091a\u094d\u091a\u093e \u091c\u0928\u094d\u092e \u0932\u0947 \u091a\u0942\u0915\u093e \u0939\u0948 \u0924\u094b \u0932\u093f\u0916\u0947 \u201c \u091c\u0928\u094d\u092e \u0928\u093e\u092e DD-MM-YY \u201d \u0914\u0930 \u092d\u0947\u0902\u091c \u0926\u0947 \u201c' + TEXTLOCAL_PHONENUMBER + '\u201d \u092a\u0930 \u0939\u093f\u0902\u0926\u0940 \u092e\u0947\u0902 \u092a\u0902\u091c\u0940\u0915\u0930\u0923 \u0915\u0947 \u0932\u093f\u090f.'
```
Comment (peterhurford) [inline modules/i18n.py]: We should just keep this to one function like the others.

File: tests/modules/test_text_processor.py [hunk 4]
```diff
@@ -236,6 +250,67 @@ def test_subscribe_twice_english(self, texting_mock, logging_info_mock, logging_
         expected_groups = ['Everyone - English', 'Text Sign Ups', 'Text Sign Ups - English']
         self.assertEqual(actual_groups, expected_groups)
 
+    @patch("logging.info")
+    @patch("modules.text_processor.Texter.send")
+    def test_pregnancy_birthdate_update_english(self, texting_mock, logging_info_mock):
+        t = TextProcessor(phone_number="1-111-1114")
+        self.assertFalse(Contact.objects.filter(name="Rose", phone_number="1-111-1114").exists())
+        first_response = t.process("JOIN ROSE 25-10-2012")
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: So the flow for pregancy sign ups is that you text `JOIN` with a birth date in the future and then text `BORN` later with an updated DOB?

File: tests/modules/test_text_reminder.py [hunk 5]
```diff
@@ -469,3 +470,67 @@ def test_remind_when_good_dont_remind_when_cancelled(self, r_mocked_send_text, t
...
+        self.assertEqual(tr.get_reminder_msg(),
+                         verify_pregnant_signup_birthdate('Hindi').format(name=u'\u0906\u0930\u0935'))
+
+    @freeze_time(FAKE_NOW)
+    def test_remind_at_four_weeks_hindi(self):
+        tr = text_reminder_object("19/6/2017", language="Hindi", preg_signup=True, preg_update=False)
+        self.assertTrue(tr.correct_date_for_reminder(weeks_after_birth=4, days_before_appointment=0))
+        self.assertTrue(tr.should_remind_today())
+        self.assertEqual(tr.get_reminder_msg(),
+                         verify_pregnant_signup_birthdate('Hindi').format(name=u'\u0906\u0930\u0935'))
```
Comment (peterhurford) [inline tests/modules/test_text_reminder.py]: We should add tests that people _without_ pregnancy _do not_ get pregnancy updates.

File: tests/modules/test_text_processor.py [hunk 6]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        self.assertTrue(t.get_contacts().exists())
+        t2 = TextProcessor(phone_number="1-111-1114")
+        second_response = t2.process("JOIN ROSE 25-11-2012")
+        self.assertTrue(t2.get_contacts().exists())
+        self.assertEqual(second_response, msg_already_sub("English"))
+        texting_mock.assert_called_with(message=second_response, phone_number="1-111-1114")
+        logging_error_mock.assert_called_with("Contact for Rose at 1-111-1114 was subscribed but already exists!")
+        contacts = Contact.objects.filter(name="Rose", phone_number="1-111-1114", preg_update=False)
+        self.assertEqual(contacts.count(), 1)
+        self.assertTrue(t2.get_contacts().exists())
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: This test doesn't actually seem to test anything about the preg update.

File: tests/modules/test_text_processor.py [hunk 7]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        t2 = TextProcessor(phone_number="1-111-1115")
+        self.assertTrue(t2.get_contacts().exists())
+        second_response = t2.process(hindi_remind() + " SANJIV 25-11-2012")
+        self.assertEqual(second_response, msg_already_sub("Hindi"))
+        texting_mock.assert_called_with(message=second_response, phone_number="1-111-1115")
+        logging_error_mock.assert_called_with("Contact for Sanjiv at 1-111-1115 was subscribed but already exists!")
+        contacts = Contact.objects.filter(name="Sanjiv", phone_number="1-111-1115", preg_update=False)
+        self.assertTrue(contacts.exists())
+        self.assertEqual(contacts.count(), 1)
+        self.assertTrue(t2.get_contacts().exists())
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: This test doesn't actually seem to test anything about the preg update.

File: tests/modules/test_text_processor.py [hunk 8]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+    def test_text_in_pregnancy_birthdate_update_english(self, texting_mock):
+        t = TextProcessor(phone_number="1-111-1114")
+        self.assertFalse(Contact.objects.filter(name="Rose", phone_number="1-111-1114").exists())
+        first_response = t.process("JOIN ROSE 25-10-2012")
+        self.assertEqual(first_response, msg_subscribe("English").format(name="Rose"))
+        texting_mock.assert_called_with(message=first_response, phone_number="1-111-1114")
+        self.assertTrue(Contact.objects.filter(name="Rose", phone_number="1-111-1114",date_of_birth=datetime(2012, 10, 25, 0, 0), preg_update=False).exists())
+        self.assertTrue(t.get_contacts().exists())
+        t2 = TextProcessor(phone_number="1-111-1114")
+        second_response = t2.process("Born ROSE 25-11-2012")
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: Should put BORN in all caps

File: tests/modules/test_text_processor.py [hunk 9]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        self.assertTrue(t2.get_contacts().exists())
+        self.assertEqual(second_response, msg_subscribe("English").format(name="Rose"))
+        texting_mock.assert_called_with(message=second_response, phone_number="1-111-1114")
+        self.assertFalse(Contact.objects.filter(name="Rose", phone_number="1-111-1114",date_of_birth=datetime(2012, 10, 25, 0, 0), preg_update=False).exists())
+        self.assertFalse(Contact.objects.filter(name="Rose", phone_number="1-111-1114",date_of_birth=datetime(2012, 11, 25, 0, 0), preg_update=False).exists())
+        self.assertTrue(Contact.objects.filter(name="Rose", phone_number="1-111-1114", date_of_birth=datetime(2012, 11, 25, 0, 0), preg_update=True).exists())
+        self.assertTrue(t2.get_contacts().exists())
+        actual_groups = [str(g) for g in t2.get_contacts().first().group_set.all()]
+        expected_groups = ['Everyone - English', 'Text Sign Ups', 'Text Sign Ups - English']
+        self.assertEqual(actual_groups, expected_groups)
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: Need to test that the preg update happens.

File: tests/modules/test_text_processor.py [hunk 10]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        self.assertTrue(t3.get_contacts().exists())
+        self.assertTrue(Contact.objects.filter(name="Tina", phone_number="910003456789", language_preference="English", date_of_birth=datetime(2017, 7, 10, 0, 0), preg_signup=True, preg_update=False).exists())
+        existing_contact_update = t3.process("Born Tina 25-07-2017")
+        self.assertEqual(existing_contact_update, msg_subscribe("English").format(name="Tina"))
+        texting_mock.assert_called_with(message=existing_contact_update, phone_number="910003456789")
+        self.assertFalse(Contact.objects.filter(name="Tina", phone_number="910003456789", language_preference="English", date_of_birth=datetime(2017, 7, 10, 0, 0), preg_signup=True, preg_update=False).exists())
+        self.assertFalse(Contact.objects.filter(name="Tina", phone_number="910003456789", language_preference="English", date_of_birth=datetime(2017, 7, 25, 0, 0), preg_signup=True, preg_update=False).exists())
+        self.assertTrue(Contact.objects.filter(name="Tina", phone_number="910003456789", language_preference="English", date_of_birth=datetime(2017, 7, 25, 0, 0), preg_signup=True, preg_update=True).exists())
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: Can we just replace L347-349 with:

```
self.assertTrue(t3.get_contacts().first().preg_update)
self.assertTrue(t3.get_contacts().first().preg_signup)
```

File: tests/modules/test_text_processor.py [hunk 11]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        self.assertTrue(t.get_contacts().exists())
+        t2 = TextProcessor(phone_number="1-111-1114")
+        second_response = t2.process(hindi_born() + " Sanjiv 25-11-2012")
+        self.assertTrue(t2.get_contacts().exists())
+        self.assertEqual(second_response, msg_subscribe("Hindi").format(name="Sanjiv"))
+        texting_mock.assert_called_with(message=second_response, phone_number="1-111-1114")
+        self.assertFalse(Contact.objects.filter(name="Sanjiv", phone_number="1-111-1114", date_of_birth=datetime(2012, 10, 25, 0, 0), preg_update=False).exists())
+        self.assertFalse(Contact.objects.filter(name="Sanjiv", phone_number="1-111-1114", date_of_birth=datetime(2012, 11, 25, 0, 0), preg_update=False).exists())
+        self.assertTrue(Contact.objects.filter(name="Sanjiv", phone_number="1-111-1114", date_of_birth=datetime(2012, 11, 25, 0, 0), preg_update=True).exists())
+        self.assertTrue(t2.get_contacts().exists())
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: see above

File: tests/modules/test_text_processor.py [hunk 12]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        self.assertTrue(t3.get_contacts().exists())
+        self.assertTrue(Contact.objects.filter(name="Sanjiv", phone_number="910003456789", language_preference="Hindi", date_of_birth=datetime(2017, 7, 10, 0, 0), preg_signup=True, preg_update=False).exists())
+        existing_contact_update = t3.process(hindi_born() + " Sanjiv 25-07-2017")
+        self.assertEqual(existing_contact_update, msg_subscribe("Hindi").format(name="Sanjiv"))
+        texting_mock.assert_called_with(message=existing_contact_update, phone_number="910003456789")
+        self.assertFalse(Contact.objects.filter(name="Sanjiv", phone_number="910003456789", language_preference="Hindi", date_of_birth=datetime(2017, 7, 10, 0, 0), preg_signup=True, preg_update=False).exists())
+        self.assertFalse(Contact.objects.filter(name="Sanjiv", phone_number="910003456789", language_preference="Hindi", date_of_birth=datetime(2017, 7, 25, 0, 0), preg_signup=True, preg_update=False).exists())
+        self.assertTrue(Contact.objects.filter(name="Sanjiv", phone_number="910003456789", language_preference="Hindi", date_of_birth=datetime(2017, 7, 25, 0, 0), preg_signup=True, preg_update=True).exists())
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: see above

File: tests/modules/test_text_processor.py [hunk 13]
```diff
@@ -260,6 +274,139 @@ def test_subscribe_twice_hindi(self, texting_mock, logging_info_mock, logging_er
...
+        self.assertTrue(Contact.objects.filter(name="Rose", phone_number="1-111-1114",date_of_birth=datetime(2012, 10, 25, 0, 0), preg_update=False).exists())
+        t2 = TextProcessor(phone_number="1-111-1114")
+        second_response = t2.process("JOIN ROSE 25-10-2012")
+        self.assertTrue(t2.get_contacts().exists())
+        self.assertEqual(second_response, msg_already_sub("English"))
+        texting_mock.assert_called_with(message=second_response, phone_number="1-111-1114")
+        logging_error_mock.assert_called_with("Contact for Rose at 1-111-1114 was subscribed but already exists!")
+        self.assertTrue(Contact.objects.filter(name="Rose", phone_number="1-111-1114").exists())
+        self.assertFalse(Contact.objects.filter(name="Rose", phone_number="1-111-1114",date_of_birth=datetime(2012, 10, 25, 0, 0), preg_update=True).exists())
+        self.assertTrue(t2.get_contacts().exists())
```
Comment (peterhurford) [inline tests/modules/test_text_processor.py]: Need to actually test something about the preg update

File: tests/modules/test_text_reminder.py [hunk 14]
```diff
@@ -469,3 +470,115 @@ def test_remind_when_good_dont_remind_when_cancelled(self, r_mocked_send_text, t
...
+    def test_dont_remind_at_two_weeks_english(self):
+        no_signup_no_update = text_reminder_object("03/7/2017", language="English", preg_signup=False, preg_update=False) # 2 weeks, 0 days ago
+        self.assertIsNone(no_signup_no_update.get_reminder_msg())
+        self.assertFalse(no_signup_no_update.should_remind_today())
+        update_no_signup = text_reminder_object("03/7/2017", language="English", preg_signup=False, preg_update=True) # 2 weeks, 0 days ago
+        self.assertIsNone(update_no_signup.get_reminder_msg())
+        self.assertFalse(update_no_signup.should_remind_today())
+        signup_and_update = text_reminder_object("03/7/2017", language="English", preg_signup=True, preg_update=True) # 2 weeks, 0 days ago
+        self.assertIsNone(signup_and_update.get_reminder_msg())
+        self.assertFalse(signup_and_update.should_remind_today())
```
Comment (peterhurford) [inline tests/modules/test_text_reminder.py]: Nice :+1:

File: tests/modules/test_text_reminder.py [hunk 15]
```diff
@@ -469,3 +470,115 @@ def test_remind_when_good_dont_remind_when_cancelled(self, r_mocked_send_text, t
...
+    @freeze_time(FAKE_NOW)
+    def test_remind_at_two_weeks_hindi(self):
+        tr = text_reminder_object("03/7/2017", language="Hindi", preg_signup=True, preg_update=False)
+        self.assertTrue(tr.correct_date_for_reminder(weeks_after_birth=2, days_before_appointment=0))
+        self.assertTrue(tr.should_remind_today())
+        self.assertEqual(tr.get_reminder_msg(),
+                         verify_pregnant_signup_birthdate('Hindi').format(name=u'\u0906\u0930\u0935'))
+
+    @freeze_time(FAKE_NOW)
+    def test_dont_remind_at_two_weeks_hindiHindi(self):
```
Comment (peterhurford) [inline tests/modules/test_text_reminder.py]: typo: `hindiHindi`

File: tests/modules/test_text_processor.py (hunk 6 above)
Comment (marcus-a-davis) [reply tests/modules/test_text_processor.py]: Yes it does. It tests whether a second signup text with "JOIN" modifies preg_update in the way that a second text with "BORN" does. Lines 283-4 establish that the contact exists and begins with preg_update False, and Line 291-2 check that after a text with "JOIN" the contact still has preg_update False. If, for example, any new text message from an existing contact updated preg_update to True, this test would not work.

File: tests/modules/test_text_processor.py (hunk 7 above)
Comment (marcus-a-davis) [reply tests/modules/test_text_processor.py]: See my answer for the English test above.

File: tests/modules/test_text_processor.py (hunk 9 above)
Comment (marcus-a-davis) [reply tests/modules/test_text_processor.py]: Line 332 tests that. Lines 330 and 331 preceding it check that it's the same same contact.

File: tests/modules/test_text_processor.py (hunk 6 above)
Comment (peterhurford) [reply tests/modules/test_text_processor.py]: Ah gotcha. I was confused. I think something like this (as I've said before)

```
self.assertTrue(t2.get_contacts().first().preg_update)
self.assertTrue(t2.get_contacts().first().preg_signup)
```

could be more explicit than searching.

File: tests/modules/test_text_processor.py (hunk 9 above)
Comment (peterhurford) [reply tests/modules/test_text_processor.py]: Gotcha.

Comment (peterhurford) [reply tests/modules/test_text_processor.py]: Wasn't used to testing this by searching for `preg_update` via `filter` but I see now.

File: tests/modules/test_text_processor.py (hunk 10 above)
Comment (marcus-a-davis) [reply tests/modules/test_text_processor.py]: While those would pass, they wouldn't be as useful. The "BORN" texts update the contact's birthday, which this filtering checks on but that alternative wouldn't test. Also, if in the future we allow multiple phone numbers to a contact these tests would all need to be rewritten.

**Pull request merged**